File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        int ty = Integer.parseInt(today.split("\\.")[0]);
        int tm = Integer.parseInt(today.split("\\.")[1]);
        int td = Integer.parseInt(today.split("\\.")[2]);
        
        Map<String, Integer> termsMap = new HashMap<>();
        for(String term : terms) {
            String[] termArr = term.split(" ");
            termsMap.put(termArr[0], Integer.parseInt(termArr[1]));
        }
        
        List<Integer> result = new ArrayList<>();
        for(int i = 0; i < privacies.length; i++) {
            String[] privacyArr = privacies[i].split(" ");
            int y = Integer.parseInt(privacyArr[0].split("\\.")[0]);
            int m = Integer.parseInt(privacyArr[0].split("\\.")[1]);
            int d = Integer.parseInt(privacyArr[0].split("\\.")[2]);
            String k = privacyArr[1];
            
            int term = termsMap.get(k);
            int ey = y + ((m + term) / 12);
            int em = (m + term) % 12;
            int ed = d - 1;
            
            if(ed == 0) {
                em -= 1;
                ed = 28;
            }
            
            if(em == 0) {
                ey -= 1;
                em = 12;
            }
            
            if(ey < ty || (ey == ty && em < tm) || (ey == ty && em == tm && ed < td)) {
                result.add(i + 1);
            }
            
        }
        
        Collections.sort(result);
        
        int[] answer = new int[result.size()];
        for(int i = 0; i < answer.length; i++) {
            answer[i] = result.get(i);
        }
        
        return answer;
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.kt
class Solution {
    fun solution(today: String, terms: Array<String>, privacies: Array<String>): IntArray {
        // 오늘 날짜 정보
        val todayArr = today.split(".")
        val todayYear = todayArr[0].toInt()
        val todayMonth = todayArr[1].toInt()
        val todayDay = todayArr[2].toInt()

        // 약관 정보 Map
        val termsMap: MutableMap<String, Int> = mutableMapOf()
        for(term in terms) {
            val termArr = term.split(" ")

            // 종류
            val kind = termArr[0]

            // 기간
            val period = termArr[1]

            termsMap[kind] = period.toInt()
        }

        val answer: MutableList<Int> = mutableListOf()
        for((idx, privacy) in privacies.withIndex()) {
            // 개인정보 날짜, 약관
            val privacyArr = privacy.split(" ")
            val privacyDate = privacyArr[0].split(".")
            val year = privacyDate[0].toInt()
            val month = privacyDate[1].toInt()
            val day = privacyDate[2].toInt()
            val kind = privacyArr[1]

            // 만료날짜
            var expiredYear = year + ((month + termsMap[kind]!!) / 12)
            var expiredMonth = (month + termsMap[kind]!!) % 12
            var expiredDay = day - 1
            
            // 만료일이 0일 경우 만료월 - 1, 만료일 28 초기화
            if(expiredDay == 0) {
                expiredMonth -= 1;
                expiredDay = 28
            }

            // 만료월이 0일 경우 만료년도 -1, 만료월 12 초기화
            if(expiredMonth == 0) {
                expiredYear -= 1;
                expiredMonth = 12
            }

            // 오늘 날짜가 만료날짜 이후인 경우
            if(todayYear > expiredYear ||
                (todayYear == expiredYear && todayMonth > expiredMonth) ||
                (todayYear == expiredYear && todayMonth == expiredMonth && todayDay > expiredDay)) {
                answer.add(idx + 1)
            }
        }

        return answer.toIntArray()
    }
}